tools/bumper/git.go

```diff
@@ -50,10 +50,126 @@ func (mgr *gitHubManager) verifyClient() error {
+	logger.Printf("Getting latest tagged from branch %s in repo %s", branch, repo)
+	tagRefs, _, err := mgr.client.Git.ListMatchingRefs(mgr.ctx, owner, repo, &github.ReferenceListOptions{Ref: "refs/tags"})
+	if err != nil {
+		return "", "", errors.Wrap(err, "Failed to get release tag refs from github client API")
+	}
+
+	branchCommits, _, err := mgr.client.Repositories.ListCommits(mgr.ctx, owner, repo, &github.CommitsListOptions{SHA: branch})
+	if err != nil {
+		return "", "", errors.Wrap(err, "Failed to get release tag refs from github client API")
+	}
+
+	// look for the first tag that belongs to the chosen branch, going over from last tag
+	for i := len(tagRefs) - 1; i >= 0; i-- {
+		tag := tagRefs[i]
+		tagName := strings.Replace(tag.GetRef(), "refs/tags/", "", 1)
+		commitShaOfTag, err := getTagCommitSha(tagName, repoDir)
+		if err != nil {
+			return "", "", errors.Wrap(err, "Failed to get commit-sha from tag name")
+		}
+
+		for _, commit := range branchCommits {
+			if commit.GetSHA() == commitShaOfTag {
+				return tagName, tag.GetObject().GetSHA(), nil
+			}
+		}
+	}
+
+	return "", "", fmt.Errorf("Error: tag not found in branch %s\n tagRefs=\n%v\nbranchCommits=\n%v\n", branch, tagRefs, branchCommits)
 }
 
+func (mgr *gitHubManager) getLatestFromBranch(repo, owner, branch, repoDir string) (string, string, error) {
+	logger.Printf("Getting Latest HEAD from branch %s in repo %s", branch, repo)
+	branchRef, _, err := mgr.client.Git.GetRef(mgr.ctx, owner, repo, "refs/heads/"+branch)
+	if err != nil {
+		return "", "", errors.Wrapf(err, "Failed to get latest HEAD ref of branch %s from github client API", branch)
+	}
+	updatedCommit := branchRef.GetObject().GetSHA()
+
+	//get virtual tag using git api
+	vtag, err := getVirtualTag(updatedCommit, repoDir)
+	if err != nil {
+		return "", "", errors.Wrapf(err, "Failed to get virtual tag of commit sha %s", updatedCommit)
+	}
+
+	return vtag, updatedCommit, nil
+}
+
+// getting commit sha from tag sha is not implemented in either github/git libraries, so we use console
+func getTagCommitSha(tagName, repoDir string) (string, error) {
+	cmd := exec.Command("git", "-C", repoDir, "rev-list", "-n", "1", tagName)
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return "", errors.Wrapf(err, "Failed to run '%s'. StdOut = %s", cmd, string(out))
+	}
+	//Trim '\n' that is given in the output string.
+	return strings.TrimSuffix(string(out), "\n"), nil
+}
+
+// getting virtual tag of non tagged commits is not implemented in either github/git libraries, so we use console
+func getVirtualTag(updatedCommit, repoDir string) (string, error) {
+	cmd := exec.Command("git", "-C", repoDir, "describe", updatedCommit, "--tags")
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return "", errors.Wrapf(err, "Failed to run git describe. StdOut = %s", string(out))
+	}
+	return strings.TrimSuffix(string(out), "\n"), nil
+}
```

Comment:
We can implement this with the golang git library so we don't depend on git version at CI.